src/sentry/types/region.py

```diff
@@ -1,8 +1,120 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable
+
+from sentry.silo import SiloMode
+
+if TYPE_CHECKING:
+    from sentry.models import Organization
+
+
+class RegionCategory(Enum):
+    GEOGRAPHIC = "GEOGRAPHIC"
+    SINGLE_TENANT = "SINGLE_TENANT"
 
 
-@dataclass
+@dataclass(frozen=True, eq=True)
 class Region:
+    """A region of the Sentry platform, hosted by a region silo."""
+
     name: str
-    subdomain: str
-    is_private: bool = False
+    """The region's unique identifier."""
+
+    id: int
+    """The region's unique numeric representation.
+
+    This number is used for composing "snowflake" IDs, and must fit inside the
+    maximum bit length specified by our snowflake ID schema.
+    """
+
+    address: str
+    """The address of the region's silo.
+
+    In a production environment, this would typically be a subdomain (e.g.,
+    "eu.sentry.io"). In a development environment, it may be useful to define
+    region addresses as a custom hostname or port number (e.g,
+    "localhost:8001").
+    """
```

Comment:
The contract for the `address` attribute is a first pass and we'll probably need to iterate on it. In production, we'll definitely need to resolve regions into the subdomain where they're being hosted, and I wanted to generalize that to arbitrary hostnames and ports. But I have no idea whether it will make sense for the strings to look like `"eu.sentry.io"`, `"https://eu.sentry.io/"`, or whatever.

Note the assumption that the production config will be supplying a value of the form `"eu.sentry.io"` rather than just `"eu"`. If it's the latter, we'll additionally need to specify the `"sentry.io"` part somewhere -- I'd prefer to keep that domain-name-building logic outside of the application code as much as possible.

Reply:
I think you're right that this might need some more thought. We will need to address non-API resources across regions as well, e.g. control silo reaching out to region silo kafka clusters. These might just be better as separate explicit configurations though

Reply:
subdomain or redirect target for default HTTPS might be useful in general however.  This address may be like, "public endpoint"-eseque?